sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java

```diff
@@ -539,37 +568,118 @@ public String toString() {
   }
 
   /**
-   * Creates a record builder with specified {@link #getSchema()}. {@link Builder#build()} will
-   * throw an {@link IllegalArgumentException} if number of fields in {@link #getSchema()} does not
-   * match the number of fields specified.
+   * Creates a row builder with specified {@link #getSchema()}. {@link Builder#build()} will throw
+   * an {@link IllegalArgumentException} if number of fields in {@link #getSchema()} does not match
+   * the number of fields specified. If any of the arguments don't match the expected types for the
+   * schema fields, {@link Builder#build()} will throw a {@link ClassCastException}.
    */
   public static Builder withSchema(Schema schema) {
     return new Builder(schema);
   }
 
+  /**
+   * Creates a row builder based on the specified row. Field values in the new row can be explicitly
+   * set using {@link ModifyingBuilder#withFieldValue}. Any values not so overridden will be the
+   * same as the values in the original row.
+   */
+  public static ModifyingBuilder fromRow(Row row) {
+    return new ModifyingBuilder(row);
+  }
+
+  /** Builder for {@link Row} that bases a row on another row. */
+  public static class ModifyingBuilder {
+    private final Row sourceRow;
+    private final Map<FieldAccessDescriptor, FieldOverride> fieldValues = Maps.newHashMap();
+
+    private ModifyingBuilder(Row sourceRow) {
+      this.sourceRow = sourceRow;
+    }
+
+    public Schema getSchema() {
+      return sourceRow.getSchema();
+    }
+
+    public ModifyingBuilder withFieldValue(String fieldName, Object value) {
```

Comment:
Wouldn't it be useful to have a `withFieldValue` with an index?

Reply:
added